customskill/skill.go

```diff
@@ -0,0 +1,90 @@
+package customskill
+
+import (
+	"encoding/json"
+	"io"
+	"reflect"
+
+	"github.com/mikeflynn/go-alexa/customskill/request"
+	"github.com/mikeflynn/go-alexa/customskill/response"
+	"github.com/pkg/errors"
+)
+
+type Skill struct {
+	ValidApplicationIDs []string
+	OnLaunch            func(*request.LaunchRequest) (*response.Envelope, error)
+	OnIntent            func(*request.IntentRequest, *request.Session) (*response.Envelope, error)
+	OnSessionEnded      func(endedRequest *request.SessionEndedRequest) error
+}
+
+var jsonMarshal = json.Marshal
+
+// Handle parses a JSON payload, calls the appropriate request handler, serializes the response, and writes it to the provided writer.
+func (s *Skill) Handle(w io.Writer, b []byte) error {
+	var (
+		resp *response.Envelope
+		err  error
+	)
+
+	m, e, err := request.BootstrapFromJSON(b)
+	if err != nil {
+		return errors.Errorf("failed to bootstrap request from JSON payload: %v", err)
+	}
+
+	if !s.applicationIDIsValid(m.Session.Application.ApplicationID) {
+		return errors.Errorf("application ID %s is not in the list of valid application IDs: %v", m.Session.Application.ApplicationID, s.ValidApplicationIDs)
+	}
+
+	switch reflect.TypeOf(e) {
+	case reflect.TypeOf(&request.LaunchRequest{}):
```

Comment:
would it be possible to do something like the type switch from here:

https://github.com/golang/go/wiki/Switch#type-switch

maybe that would eliminate the need for the reflect package?

Reply:
not sure. Let me look into it and get back to you